python/ray/data/datasource/file_based_datasource.py

```diff
@@ -234,6 +240,26 @@ def _read_file(self, f: "pyarrow.NativeFile", path: str, **reader_args) -> Block
             "Subclasses of FileBasedDatasource must implement _read_file()."
         )
 
+    def _convert_block_to_tabular_block(
+        self, block: Block
+    ) -> Union["pyarrow.Table", "pd.DataFrame"]:
+        """Convert block returned by `_read_file` or `_read_stream` to a tabular block.
+
+        If your `_read_file` or `_read_stream` implementation returns a list,
+        then you need to implement this method. Otherwise, `FileBasedDatasource` won't
+        be able to include partition data.
+        """
+        import pandas as pd
+        import pyarrow as pa
+
+        if isinstance(block, (pd.DataFrame, pa.Table)):
+            return block
```

Comment:
> This would make the Datasource-level more pandas/Arrow-agnostic, and would centralize everything in the BlockAccessor level, which other datasources would automatically benefit from.

Agreed. Seems like this'd be a worthwhile refactor in the future.

Reply:
Created issue to track: https://github.com/ray-project/ray/issues/28866